src/passes/SafeHeap.cpp

```diff
@@ -0,0 +1,316 @@
+          for (auto isAtomic : { true, false }) {
+            store.isAtomic = isAtomic;
+            if (store.isAtomic && align != bytes) continue;
+            if (store.isAtomic && !module->memory.shared) continue;
+            addStoreFunc(&store, module);
+          }
+        }
+      }
+    }
+  }
+
+  void addLoadFunc(Load* curr, Module* module) {
+    auto* func = new Function;
+    func->name = getLoadName(curr);
+    func->params.push_back(i32); // pointer
+    func->params.push_back(i32); // offset
+    func->vars.push_back(i32); // pointer + offset
+    func->result = curr->type;
+    Builder builder(*module);
+    auto* block = builder.makeBlock();
+    block->list.push_back(
+      builder.makeSetLocal(
+        2,
+        builder.makeBinary(
+          AddInt32,
+          builder.makeGetLocal(0, i32),
+          builder.makeGetLocal(1, i32)
+        )
+      )
+    );
+    // check for reading past valid memory: if pointer + offset + bytes
+    block->list.push_back(
+      makeBoundsCheck(curr->type, builder, 2)
+    );
+    // check proper alignment
+    if (curr->align > 1) {
+      block->list.push_back(
+        makeAlignCheck(curr->align, builder, 2)
+      );
+    }
+    // do the load
+    auto* load = module->allocator.alloc<Load>();
+    *load = *curr; // basically the same as the template we are given!
+    load->ptr = builder.makeGetLocal(2, i32);
+    block->list.push_back(load);
+    block->finalize(curr->type);
+    func->body = block;
+    module->addFunction(func);
+  }
+
+  void addStoreFunc(Store* curr, Module* module) {
+    auto* func = new Function;
+    func->name = getStoreName(curr);
+    func->params.push_back(i32); // pointer
+    func->params.push_back(i32); // offset
+    func->params.push_back(curr->valueType); // value
+    func->vars.push_back(i32); // pointer + offset
+    func->result = none;
+    Builder builder(*module);
+    auto* block = builder.makeBlock();
+    block->list.push_back(
+      builder.makeSetLocal(
+        3,
+        builder.makeBinary(
+          AddInt32,
+          builder.makeGetLocal(0, i32),
+          builder.makeGetLocal(1, i32)
+        )
+      )
+    );
+    // check for reading past valid memory: if pointer + offset + bytes
+    block->list.push_back(
+      makeBoundsCheck(curr->valueType, builder, 3)
+    );
+    // check proper alignment
+    if (curr->align > 1) {
+      block->list.push_back(
+        makeAlignCheck(curr->align, builder, 3)
+      );
+    }
+    // do the store
+    auto* store = module->allocator.alloc<Store>();
+    *store = *curr; // basically the same as the template we are given!
+    store->ptr = builder.makeGetLocal(3, i32);
+    store->value = builder.makeGetLocal(2, curr->valueType);
+    block->list.push_back(store);
+    block->finalize(none);
+    func->body = block;
+    module->addFunction(func);
+  }
+
+  Expression* makeAlignCheck(Address align, Builder& builder, Index local) {
+    return builder.makeIf(
```

Comment:
I guess this is supposed to fault if the load is misaligned, which is right for asm.js, but misaligned loads are allowed in wasm. So, by default we might want to fault (or warn) anyway, since there could be performance implications, but we might want different handling for wasm. Also misalignment is not allowed for atomics, so we might want to handle those differently too.

Reply:
Yeah, maybe there should be an option here. But I think to start with that faulting on incorrectly-aligned load/store is best. Also faulting keeps it closer to asm.js which helps with debugging in some cases. So I tend to think an option to ignore this would make more sense farther in the future.